File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/find_first_and_last_position_of_element_in_sorted_array/FindFirstandLastPositionofElementinSortedArray.java
package main.java.com.jiang.leetcode100.find_first_and_last_position_of_element_in_sorted_array;

import java.util.Scanner;

class Solution {
    public int[] searchRange(int[] nums, int target) {
        int n = nums.length;
        int[] result = new int[]{-1, -1};
        if(n <= 0) return result;

        int left = 0, right = n - 1;
        
        while(left <= right) {
            int mid = (left + right) >> 1;
            if(nums[mid] < target) left = mid + 1;
            else right = mid - 1;
        }
        if(left < n && nums[left] == target) result[0] = left;
        else return result;
        right = n - 1;
        while(left <= right) {
            int mid = (left + right) >> 1;
            if(nums[mid] <= target) left = mid + 1;
            else right = mid - 1;
        }
        if(nums[right] == target) result[1] = right;

        return result;
    }
}

public class FindFirstandLastPositionofElementinSortedArray {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        int target = scanner.nextInt();
        scanner.close();;
        String[] nums_str = str.split(" ");
        int[] nums = new int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) {
            nums[i] = Integer.parseInt(nums_str[i]);
        }

        int[] result = new Solution().searchRange(nums, target);
        System.out.println(result[0] + " " + result[1]);
    }
}

File: Leetcode100/leetcode100_python/FindFirstandLastPositionofElementinSortedArray.py
from typing import List

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def binarySearch(left, right, target):
            while left <= right:
                mid = (left + right) // 2
                if nums[mid] < target:
                    left = mid + 1
                else:
                    right = mid - 1
            return left
        first = binarySearch(0, len(nums) - 1, target)
        if first >= len(nums) or nums[first] != target:
            return [-1, -1]
        second = binarySearch(first, len(nums) - 1, target + 1)
        return [first, second - 1]

if __name__ == "__main__":
    nums = [int(num) for num in input().split()]
    target = int(input())
    solution = Solution()
    print(solution.searchRange(nums, target))